src/envs/playroom.py

```diff
@@ -0,0 +1,507 @@
+        block_pose = np.array([state.get(block, "pose_x"),
+                               state.get(block, "pose_y"),
+                               state.get(block, "pose_z")])
+        rotation = state.get(robot, "rotation")
+        arr = np.r_[block_pose+params, rotation, 0.0].astype(np.float32)
+        arr = np.clip(arr, self.action_space.low, self.action_space.high)
+        return Action(arr)
+
+    def _Stack_policy(self, state: State, memory: Dict,
+                      objects: Sequence[Object], params: Array) -> Action:
+        # Differs from blocks because need robot rotation
+        del memory  # unused
+        robot, block = objects
+        block_pose = np.array([state.get(block, "pose_x"),
+                               state.get(block, "pose_y"),
+                               state.get(block, "pose_z")])
+        rotation = state.get(robot, "rotation")
+        arr = np.r_[block_pose+params, rotation, 1.0].astype(np.float32)
+        arr = np.clip(arr, self.action_space.low, self.action_space.high)
+        return Action(arr)
+
+    def _PutOnTable_policy(self, state: State, memory: Dict,
+                           objects: Sequence[Object], params: Array) -> Action:
+        # Differs from blocks because need robot rotation, table bounds
+        del memory  # unused
+        robot, = objects
+        # Un-normalize parameters to actual table coordinates
+        x_norm, y_norm = params
+        x = self.table_x_lb + (self.table_x_ub - self.table_x_lb) * x_norm
+        y = self.table_y_lb + (self.table_y_ub - self.table_y_lb) * y_norm
+        z = self.table_height + 0.5*self.block_size
+        rotation = state.get(robot, "rotation")
+        arr = np.array([x, y, z, rotation, 1.0], dtype=np.float32)
+        arr = np.clip(arr, self.action_space.low, self.action_space.high)
+        return Action(arr)
+
+    def _robot_is_facing_dial(self, state: State, action: Action) -> bool:
+        x, y, _, rotation, _ = action.arr
+        dial_x = state.get(self._dial, "pose_x")
+        dial_y = state.get(self._dial, "pose_y")
+        theta = np.arctan2(dial_y-y, dial_x-x)
+        if np.pi*3/4 >= theta > np.pi/4:  # N
+            if 1.5 >= rotation > 0.5:
+                return True
+        elif np.pi/4 >= theta > -np.pi/4:  # E
+            if 0.5 >= rotation > -0.5:
+                return True
+        elif -np.pi/4 >= theta > -np.pi*3/4:  # S
+            if -0.5 >= rotation > -1.5:
+                return True
+        else:  # W
+            if rotation > 1.5 or rotation <= -1.5:
+                return True
+        return False
+
+    def _get_door_next_to(self, state: State) -> Object:
+        # cannot be next to multiple doors at once
+        for door in self._doors:
+            if self._NextToDoor_holds(state, (self._robot, door)):
+                return door
+        # usage should ensure this is never reached
+        raise RuntimeError("Robot not next to any door")  # pragma: no cover
```

Comment:
Is it hard to write a test that reaches this line, just by setting up an impossible state and calling this function directly?

Reply:
e.g.
env = PlayroomEnv()
state = env.get_train_tasks()[0].init
(change state in some way)
with pytest.raises(RuntimeError):
  env._get_door_next_to(state)  # pylint: disable=protected-access